### PR TITLE
Feature FEELPP_HAS

### PR DESCRIPTION
**Pull request by prudhomm**
- [ X Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?
- [ ] Have you written Doxygen comments in your contribution ?

-----

it implements #1218

### CODE REVIEW

**Comment by prudhomm** [comment]
@vincentchabannes please have a look and let me know. I believe it solves your concern and moreover it is simple, clean, we do not have to maintain a set of global variable.
of course this needs to be tested on a few configurations to make sure I haven't left out anything

**Comment by prudhomm** [comment]
@vincentchabannes as you can see in the last commit I added a function that extracts and generates the cmake variables `FEELPP_HAS_<>`

**Comment by prudhomm** [comment]
@vincentchabannes now you have to review the PR ;)

File: testsuite/CMakeLists.txt [hunk 1]
```diff
@@ -28,8 +28,11 @@ set(PROJECT_SHORTNAME "test")
 if ( ${CMAKE_SOURCE_DIR}/ STREQUAL ${CMAKE_CURRENT_SOURCE_DIR} )
   find_package(Feel++ PATHS $ENV{FEELPP_DIR}/share/feelpp/feel/cmake/modules /usr/share/feelpp/feel/cmake/modules /usr/local/share/feelpp/feel/cmake/modules )
   if(NOT FEELPP_FOUND)
-      message(FATAL_ERROR "Feel++ was not found on your system. Make sure to install it and specify the FEELPP_DIR to reference the installation directory.")
+    message(FATAL_ERROR "Feel++ was not found on your system. Make sure to install it and specify the FEELPP_DIR to reference the installation directory.")
   endif()
+else()
+  feelpp_set_options(Feelpp::feelpp_contrib ${PROJECT_NAME})
+  feelpp_set_options(Feelpp::feelpp ${PROJECT_NAME})
```
Comment (prudhomm) [inline testsuite/CMakeLists.txt]: @vincentchabannes perhaps `feelpp_set_cmake_options` would be a better name. it is probably too complicated to have two lines and we should have only one line

Comment (vincentchabannes) [reply testsuite/CMakeLists.txt]: is it really necessary to use Feelpp ::feelpp_contrib, the definitions are not propagated in Feelpp::feelpp?

Comment (prudhomm) [reply testsuite/CMakeLists.txt]: it is propagated via `target_link_libraries` as a dependencies but the variables are not set in feelpp. if we do that, then we have duplicates

Comment (prudhomm) [reply testsuite/CMakeLists.txt]: what could be done is hide the call to `feelpp_contrib` and `feelpp` inside a `feelpp_set_cmake_options` or something like that

**Comment by prudhomm** [comment]
the result of the testsuite:
```
98% tests passed, 5 tests failed out of 206

Label Time Summary:
testalg              =  23.51 sec*proc (16 tests)
testcore             =  26.06 sec*proc (22 tests)
testcrb              =  82.93 sec*proc (9 tests)
testdiscr            = 304.65 sec*proc (64 tests)
testfilters          =  32.59 sec*proc (7 tests)
testfit              =   2.42 sec*proc (2 tests)
testintegration      =  14.77 sec*proc (7 tests)
testinterpolation    =  61.21 sec*proc (16 tests)
testleaks            =  12.77 sec*proc (8 tests)
testls               =   3.31 sec*proc (2 tests)
testmaterial         =   1.40 sec*proc (1 test)
testmath             =   0.80 sec*proc (1 test)
testmesh             =  30.53 sec*proc (13 tests)
testmodels           =   2.61 sec*proc (1 test)
testopt              =   0.00 sec*proc (1 test)
testpde              =   6.67 sec*proc (2 tests)
testpoly             =  21.72 sec*proc (9 tests)
testts               =   7.06 sec*proc (5 tests)
testvf               =  49.16 sec*proc (19 tests)

Total Test time (real) = 349.50 sec

The following tests did not run:
	104 - feelpp_test_space_0d-np-1 (Skipped)
	169 - feelpp_test_bdf3-np-1 (Skipped)
	175 - feelpp_test_laplacianv-np-1 (Skipped)
	176 - feelpp_test_laplacian-np-1 (Skipped)
	177 - feelpp_test_laplaciant-np-1 (Skipped)

The following tests FAILED:
	 21 - feelpp_test_serialization-np-1 (Failed)
	 38 - feelpp_test_eigenmode-np-1 (Failed)
	164 - feelpp_test_db-np-1 (Failed)
	179 - feelpp_test_inner_curl-np-1 (Not Run)
	199 - feelpp_test_nlopt-np-1 (Not Run)
Errors while running CTest
```
I thbink we are good to go

**Comment by vincentchabannes** [comment]
There is problem with testsuite, only feelpp_test_inner_curl should fails

**Comment by vincentchabannes** [comment]
The results of my testsuite : 
```
The following tests did not run:
        104 - feelpp_test_space_0d-np-1 (Skipped)
        169 - feelpp_test_bdf3-np-1 (Skipped)
        175 - feelpp_test_laplacianv-np-1 (Skipped)
        176 - feelpp_test_laplacian-np-1 (Skipped)
        177 - feelpp_test_laplaciant-np-1 (Skipped)

The following tests FAILED:
         21 - feelpp_test_serialization-np-1 (Failed)
         36 - feelpp_test_prepost_solve-np-1 (Failed)
         38 - feelpp_test_eigenmode-np-1 (Failed)
         60 - feelpp_test_integrateQuadra-np-1 (Failed)
         70 - feelpp_test_modelproperties-np-1 (Failed)
         76 - feelpp_test_eval_at_point-np-1 (SEGFAULT)
         84 - feelpp_test_darcy_on_2d-np-1 (Failed)
         85 - feelpp_test_darcy_on_3d-np-1 (Failed)
         89 - feelpp_test_integration_ginac-np-1 (SEGFAULT)
        114 - feelpp_test_continuity_h1-np-1 (Failed)
        115 - feelpp_test_continuity_hdiv-np-1 (Failed)
        135 - feelpp_test_element_component-np-1 (Failed)
        142 - feelpp_test_mortar2-np-1 (Failed)
        164 - feelpp_test_db-np-1 (Failed)
        168 - feelpp_test_bdf2-np-1 (Failed)
        172 - feelpp_test_ginac-np-1 (Failed)
        173 - feelpp_test_inner-np-1 (Failed)
        178 - feelpp_test_msi-np-1 (Failed)
        179 - feelpp_test_inner_curl-np-1 (Not Run)
        184 - feelpp_test_pow-np-1 (Failed)
        197 - feelpp_test_solve_leak-np-1 (SEGFAULT)
        199 - feelpp_test_nlopt-np-1 (Not Run)
        202 - feelpp_test_ams-np-1 (Failed)
        203 - feelpp_test_saddle-np-1 (Failed)
```

**Comment by vincentchabannes** [comment]
@prudhomm My problems appear when an expression is compiled with Ginac. I don't understand what's going on.
For example with qs_laplacian : 
```
./feelpp/quickstart/feelpp_qs_laplacian_2d --functions.f=x:x
[ Starting Feel++ ] application qs_laplacian version 0.106.0-beta.20 date 2018-Sep-25
 . qs_laplacian files are stored in /home/u2/chabannes/feel/qs_laplacian/np_1
 .. logfiles :/home/u2/chabannes/feel/qs_laplacian/np_1/logs
[loadMesh] no file name or unrecognized extension provided
[loadMesh] automatically generating amesh from gmsh.domain.shape in format geo+msh: "hypercube".geo
[loadMesh] Time : 0.0950596s
*** Aborted at 1537891774 (unix time) try "date -d @1537891774" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGSEGV (@0x0) received by PID 73165 (TID 0x7f5acc68f3c0) from PID 0; stack trace: ***
    @     0x7f5aaae4c390 (unknown)
    @     0x7f5a9ec5de2d boost::filesystem::detail::status()
    @     0x7f5a9f2ac247 GiNaC::excompiler::compile_src_file()
    @     0x7f5a9f2ab32e GiNaC::compile_ex()
    @     0x7f5acbc4e53a Feel::vf::detail::ginacBuildLibrary()
    @           0x518af7 _ZN4Feel2vf9GinacExVFILi2ENS0_11SymbolsExprIJEEEEC2ERKN5GiNaC2exERKSt6vectorINS5_6symbolESaISA_EERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESK_RKNS_9WorldCommESM_RKS3_
    @           0x5181a0 _ZN4Feel2vf4exprILi2EEENS0_4ExprINS0_9GinacExVFIXT_ENS0_11SymbolsExprIJEEEEEEERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESD_RKNS_9WorldCommESF_
    @           0x508a68 main
    @     0x7f5a9d815830 __libc_start_main
    @           0x5081d9 _start
    @                0x0 (unknown)
Segmentation fault (core dumped)
```
But if I  don't compile the expression, it's ok : 
```
./feelpp/quickstart/feelpp_qs_laplacian_2d --functions.f=1  
[ Starting Feel++ ] application qs_laplacian version 0.106.0-beta.20 date 2018-Sep-25
 . qs_laplacian files are stored in /home/u2/chabannes/feel/qs_laplacian/np_1
 .. logfiles :/home/u2/chabannes/feel/qs_laplacian/np_1/logs
[loadMesh] no file name or unrecognized extension provided
[loadMesh] automatically generating amesh from gmsh.domain.shape in format geo+msh: "hypercube".geo
[loadMesh] Time : 0.0870925s
[Vh] Time : 0.00651369s
[l] Time : 0.0175077s
 [a] Time : 0.0182717s
[a] Time : 0.0252646s
[a.solve] Time : 0.0144971s
 [Exporter] Time : 0.0263525s
[env] Time : 0.187135s
[ Stopping Feel++ ] application qs_laplacian execution time 0.187135s
```
any idea?

**Comment by prudhomm** [comment]
Yes will fix tonight

**Comment by vincentchabannes** [comment]
Also I noticed that dependencies of boost is not done in feelpp but in contrib.
If correct, I think we need to check if we use a minimal version (only boost and mpi) by deactivated all module, I am not sure that boost is correctly included/linked

**Comment by vincentchabannes** [comment]
@prudhomm result of my testsuite : 
```
99% tests passed, 3 tests failed out of 205

Label Time Summary:
testalg              =  44.58 sec (16 tests)
testcore             =  52.41 sec (22 tests)
testcrb              = 108.54 sec (9 tests)
testdiscr            = 402.21 sec (64 tests)
testfilters          =  36.81 sec (7 tests)
testfit              =   5.63 sec (2 tests)
testintegration      =  25.20 sec (7 tests)
testinterpolation    =  87.68 sec (16 tests)
testleaks            =  22.95 sec (8 tests)
testls               =   5.66 sec (2 tests)
testmaterial         =   2.03 sec (1 test)
testmath             =   2.02 sec (1 test)
testmesh             =  45.80 sec (13 tests)
testmodels           =   5.93 sec (1 test)
testopt              =   0.00 sec (1 test)
testpde              =   9.56 sec (2 tests)
testpoly             =  38.50 sec (9 tests)
testts               =  14.01 sec (5 tests)
testvf               =  79.18 sec (19 tests)

Total Test time (real) = 990.05 sec

The following tests did not run:
        104 - feelpp_test_space_0d-np-1 (Skipped)
        169 - feelpp_test_bdf3-np-1 (Skipped)
        175 - feelpp_test_laplacianv-np-1 (Skipped)
        176 - feelpp_test_laplacian-np-1 (Skipped)
        177 - feelpp_test_laplaciant-np-1 (Skipped)

The following tests FAILED:
         38 - feelpp_test_eigenmode-np-1 (Failed)
        179 - feelpp_test_inner_curl-np-1 (Not Run)
        199 - feelpp_test_nlopt-np-1 (Not Run)
```

It's almost good, still the nlopt test and to check with differents configurations (enable/disable some dependencies).

**Comment by vincentchabannes** [comment]
@prudhomm There is no parallel test, something is still wrong in cmake files

**Comment by prudhomm** [comment]
it is because `NProcs2` is not defined, I am adding to toplevel

**Comment by vincentchabannes** [comment]
@prudhomm nlopt should be fix now, see my previous commit.
But I don't see the quickstart testsuite in buildkite, is it normal?

**Comment by prudhomm** [comment]
@vincentchabannes I think that is safe to merge in the current state of things, no ?

**Comment by prudhomm** [comment]
there might be a few minor cleanups until the final release

**Pull request merged**